### DISCUSSION TITLE
Chunk Hierarchy Design Discussion

### DISCUSSION THREAD

**Discussion by balacij** [JacquesCarette/Drasil#3773] [General]
Let's use this thread to discuss the chunk hierarchy in Drasil. New comments = new "discussion threads." Alternatively, we can create a new "Chunk Design" discussions category and create individual discussions related to the chunk design, there.

**Comment by balacij** [comment]
@JacquesCarette @smiths I've shared @NoahCardoso and I's (new) drawio files via Gmail (I can share it to everyone else if they can send us their Gmail emails to either of us on Discord). We focused on the top-most pieces right now, incorporating a few ideas from past discussions we've had.
1. OldChunk.drawio is really the current chunk design.
2. AttemptAtChunksRev2.drawio is a proposal for how we should be defining chunks with Drasil. Only the top half of this file is actually edited right now. We have some writing on the LHS about what changes were actually made and some notes floating around that describe the intentions of the proposed chunks at the "top" of the hierarchy, focusing on how we define abstract and concrete ideas, and how we might want to associate symbols with some ideas. Admittedly, the both parts are rather large changes, so trying to tackle both at the same time might have been too ambitious.

Please let us know if the changes to the top of the hierarchy make sense and what you think about them. Our examples remain scarce, but I think we need to build out a few more chunk types to better get the point across (of the design) and showcase how chunks from our existing case studies would fit in the proposed design.

**Reply by smiths** [reply]
Nice to see @balacij and @NoahCardoso tackling this.  I've added this to our next Drasil meeting agenda.

How does this design relate to what @JacquesCarette presented in our [chunk redesign meetings](https://github.com/JacquesCarette/Drasil/wiki/Chunk-Redesign-2024)?  The example helps, but I'll probably need many more examples before I fully get it.  :smile:

**Reply by samm82** [reply]
Is there a way to share the drawio files through the link itself? That could just be posted (and probably pinned for easy reference) in our Drasil discord if so

**Reply by smiths** [reply]
The XML file can be exported from draw.io and put under version control.  The user would still need to open the file in draw.io, but then we have the history of the evolution of the diagram (if we need it).  Might be worth doing.

**Comment by JacquesCarette** [comment]
[Some relevant comments](https://github.com/JacquesCarette/Drasil/issues/3742#issuecomment-2130061240) from Sam.

**Comment by JacquesCarette** [comment]
This seems to be nicely coming together. Here is my current take, based on our meetings and the diagrams.

**Definition**: A *chunk* is a fundamental piece of information that we need to represent. A *type of chunk* classifies the kind of information that it is, recognizing that we have different needs.

So here goes some *chunk types*:
- **Idea**. In math, this would basically be a 'type'. (abstract type?) It has an internal name (UID) and an external name (NounPhrase).
- **IdeaDef**. In math, this would be a 'defined type'. It has all the data of an Idea plus also a definition. I don't think it should be implemented via nesting!
- **ConcreteIdea** (?). This is an **Idea** with a concrete list of **Properties**, i.e. queries we can perform. Think of 'pair' with 'fst' and 'snd' as known queries.
- **Thing** (name can change - I was tempted by both **Object** (too overloaded) and **ConcreteConcept** (urgh)). This should be an instance of an **Idea**. In math, this would be an actual number, or an actual Vector, or a function or ...  The important part is that a **Thing** can have properties.
- **Variable**. Same name in math -- its stands for a value (which I called **Thing** above) of any kind of **Idea** (i.e. could be **IdeaDef** and **ConcreteIdea** too).

We need to have some **Properties** too:
- **Abbreviation**: a property of any kind of idea
- **Notation**: a property of a **Variable**
- **Unit**: a property for both **Thing** and **Variable**
We will want some higher-level information gathering means too, like **Theory** and **Scope**. But I don't see either of those as being chunks themselves?
- **Uncertainty**: a property of certain kinds of **Variables**, i.e. being able to associate an uncertainty to a variable would depend on that variable itself having some other property

A decent test of any new design: take ~10 chunks from Projectile (pick them to be of a rather different nature from each other) and explain all of them using the new design.
This is incomplete - we need to be able to see where all of the old chunks would fit.

**Reply by smiths** [reply]
I like this description.  It helps with forming a mental model.  I agree that looking at about 10 chunks in Projectile would help me deepen my understanding.  I know we are likely trying to avoid nesting, but I'd like to also talk about the relationships between chunks.  For instance, IdeaDef doesn't wrap around Idea, but does it point to an Idea, or are IdeaDef and Idea not related?
